proposals/0000-extend-array-primitives.md

```diff
@@ -0,0 +1,80 @@
+---
+author: buggymcbugfix
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/334>).
+
+# RFC: Expand arrays APIs with efficient primops
+
+## Motivation
+
+Operations that allocate new arrays (`Array#`, `MutableArray#`, `SmallArray#`,
+`SmallMutableArray#`) must ensure that the array never contains garbage values
+during safe points (when the GC runs); otherwise the GC would interpret these as
+pointers and try to follow them and Undefined Behaviour would ensue. Thus, for
+safety, many library-defined array operations have to conservatively initialise
+array slots at the cost of performance.
+
+We propose to expand the arrays API in
+[GHC.Exts](https://hackage.haskell.org/package/base-4.14.0.0/docs/GHC-Exts.html)
+by picking the low hanging fruit of missing primitives which would improve the
+performance of key libraries, such as
+[unordered-containers](https://github.com/tibbe/unordered-containers/tree/d0b78afc028ef307a3cdfe1dd95feaf4ec47f195).
+
+## Proposed Change Specification
+
+Below is the current list of proposed new primops, in order of priority.
+
+- `insert# :: array# a -> Int# -> a -> array# a`
```

Comment:
Please state the the expected semantics for the primops in the proposal itself.

Reply:
Oops, I never defined what `concat` is meant to do. But for the other ones, is there any value in copying and pasting the `unordered-containers` code?